File: exec/src/mcp/service.ts
/**
 * facade 业务逻辑：上下文路由、桥调用、产物 URL 签名。
 * 移植自 `sandbox/mcp/service.py`。
 *
 * ## Model Experience
 * 每个工具结果都以 `context_id` 开头回显，模型据此把后续调用绑到同一个工作区
 * ——这是它唯一能表达"接着刚才那个工作区做"的手段。上限（代码长度、命令长度、
 * 文件大小、超时）在**发桥之前**就判定，所以超限的失败不消耗执行面时间，
 * 拒绝文案是固定短句，对 KV cache 友好。
 *
 * ## Known Limitations and Deferred Work
 * - 下载 URL 的签名只绑 `artifact_id` 与过期时间，不绑调用者：拿到 URL 的人在
 *   TTL 内都能下载。与 Python 版一致；要绑调用者需要 facade 侧有身份概念。
 */
import { createHmac, timingSafeEqual } from 'node:crypto';
import { SandboxBridgeClient, SandboxBridgeError } from './bridge-client.js';
import { ContextStore, ContextStoreError, type ContextRecord } from './context-store.js';
import type { McpSettings } from './settings.js';
import { newUlid } from './ulid.js';
import { validateRuntime } from './settings.js';

/** 经 MCP 协议返回给调用方的安全消息。 */
export class McpFacadeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'McpFacadeError';
  }
}

type Json = Record<string, unknown>;

function b64(data: Buffer): string {
  return data.toString('base64url');
}

function unb64(value: string): Buffer {
  return Buffer.from(value, 'base64url');
}

const MIME_BY_EXT: Readonly<Record<string, string>> = {
  '.txt': 'text/plain',
  '.md': 'text/markdown',
  '.html': 'text/html',
  '.htm': 'text/html',
  '.css': 'text/css',
  '.csv': 'text/csv',
  '.json': 'application/json',
  '.js': 'text/javascript',
  '.xml': 'application/xml',
  '.pdf': 'application/pdf',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.webp': 'image/webp',
  '.zip': 'application/zip',
  '.gz': 'application/gzip',
  '.tar': 'application/x-tar',
  '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  '.xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  '.pptx': 'application/vnd.openxmlformats-officedocument.presentationml.presentation',
};

function guessMimeType(name: string): string | null {
  const dot = name.lastIndexOf('.');
  if (dot < 0) return null;
  return MIME_BY_EXT[name.slice(dot).toLowerCase()] ?? null;
}

export class McpFacadeService {
  readonly #settings: McpSettings;
  readonly #contextStore: ContextStore;
  readonly #bridge: SandboxBridgeClient;

  constructor(settings: McpSettings, contextStore: ContextStore, bridge: SandboxBridgeClient) {
    this.#settings = settings;
    this.#contextStore = contextStore;
    this.#bridge = bridge;
  }

  async start(): Promise<void> {
    validateRuntime(this.#settings);
    await this.#contextStore.start();
    this.#bridge.start();
  }

  async close(): Promise<void> {
    this.#bridge.close();
    await this.#contextStore.close();
  }

  static #contextPayload(record: ContextRecord): Json {
    return {
      sandbox_session_id: record.sandboxSessionId,
      workspace_id: record.workspaceId,
    };
  }

  async #resolve(contextId: string | null | undefined): Promise<ContextRecord> {
    try {
      return await this.#contextStore.resolve(contextId, (record) =>
        this.#bridge.post('/internal/mcp/v1/context/ensure', McpFacadeService.#contextPayload(record)),
      );
    } catch (error) {
      if (error instanceof ContextStoreError || error instanceof SandboxBridgeError) {
        throw new McpFacadeError(error.message);
      }
      throw error;
    }
  }

  async #call(
    path: string,
    contextId: string | null | undefined,
    payload: Json,
  ): Promise<[ContextRecord, Json]> {
    const record = await this.#resolve(contextId);
    try {
      const data = await this.#bridge.post(path, {
        ...McpFacadeService.#contextPayload(record),
        ...payload,
      });
      await this.#contextStore.touch(record);
      return [record, data];
    } catch (error) {
      if (error instanceof ContextStoreError || error instanceof SandboxBridgeError) {
        throw new McpFacadeError(error.message);
      }
      throw error;
    }
  }

  #checkTimeout(timeoutSeconds: number): void {
    if (!(timeoutSeconds >= 1 && timeoutSeconds <= this.#settings.maxTimeoutSeconds)) {
      throw new McpFacadeError('timeout_seconds exceeds MCP limit');
    }
  }

  async executePython(input: {
    contextId?: string | null;
    code: string;
    timeoutSeconds?: number;
  }): Promise<Json> {
    const timeoutSeconds = input.timeoutSeconds ?? 120;
    if (input.code.length > this.#settings.maxCodeLength) {
      throw new McpFacadeError('code exceeds MCP size limit');
    }
    this.#checkTimeout(timeoutSeconds);
    const [record, data] = await this.#call('/internal/mcp/v1/python/execute', input.contextId, {
      code: input.code,
      timeout_seconds: timeoutSeconds,
    });
    return { context_id: record.contextId, ...data };
  }

  async executeShell(input: {
    contextId?: string | null;
    command: string;
    timeoutSeconds?: number;
  }): Promise<Json> {
    const timeoutSeconds = input.timeoutSeconds ?? 120;
    if (input.command.length > this.#settings.maxCommandLength) {
      throw new McpFacadeError('command exceeds MCP size limit');
    }
    this.#checkTimeout(timeoutSeconds);
    const [record, data] = await this.#call('/internal/mcp/v1/shell/execute', input.contextId, {
      command: input.command,
      timeout_seconds: timeoutSeconds,
    });
    return { context_id: record.contextId, ...data };
  }

  async fileWrite(input: {
    contextId?: string | null;
    path: string;
    content: string;
    mode?: string;
  }): Promise<Json> {
    const mode = input.mode ?? 'overwrite';
    if (mode !== 'overwrite' && mode !== 'append') {
      throw new McpFacadeError('mode must be overwrite or append');
    }
    if (Buffer.byteLength(input.content, 'utf8') > this.#settings.maxFileSizeBytes) {
      throw new McpFacadeError('content exceeds MCP file size limit');
    }
    const [record, data] = await this.#call('/internal/mcp/v1/files/write', input.contextId, {
      path: input.path,
      content: input.content,
      mode,
    });
    return { context_id: record.contextId, ...data };
  }

  async fileRead(input: {
    contextId?: string | null;
    path: string;
    offset?: number | null;
    limit?: number | null;
  }): Promise<Json> {
    const [record, data] = await this.#call('/internal/mcp/v1/files/read', input.contextId, {
      path: input.path,
      offset: input.offset ?? null,
      limit: input.limit ?? null,
    });
    return { context_id: record.contextId, ...data };
  }

  async fileList(input: {
    contextId?: string | null;
    path?: string;
    depth?: number;
  }): Promise<Json> {
    const depth = input.depth ?? 1;
    if (!(depth >= 0 && depth <= 5)) throw new McpFacadeError('depth must be in 0..5');
    const [record, data] = await this.#call('/internal/mcp/v1/files/list', input.contextId, {
      path: input.path ?? '.',
      depth,
    });
    return { context_id: record.contextId, ...data };
  }

  #signArtifact(artifactId: string, expiresAt: number): string {
    const body = b64(Buffer.from(JSON.stringify({ artifact_id: artifactId, exp: expiresAt })));
    const signature = createHmac('sha256', this.#settings.downloadSecret).update(body).digest();
    return `${body}.${b64(signature)}`;
  }

  verifyArtifactToken(artifactId: string, token: string): boolean {
    try {
      const dot = token.indexOf('.');
      if (dot < 0) return false;
      const body = token.slice(0, dot);
      const signature = token.slice(dot + 1);
      const expected = createHmac('sha256', this.#settings.downloadSecret).update(body).digest();
      const provided = unb64(signature);
      if (provided.length !== expected.length) return false;
      if (!timingSafeEqual(provided, expected)) return false;
      const payload: unknown = JSON.parse(unb64(body).toString('utf8'));
      if (payload === null || typeof payload !== 'object' || Array.isArray(payload)) return false;
      const record = payload as Record<string, unknown>;
      return (
        record['artifact_id'] === artifactId && Number(record['exp'] ?? 0) >= Math.floor(Date.now() / 1000)
      );
    } catch {
      return false;
    }
  }

  async artifactSubmit(input: {
    contextId?: string | null;
    sourcePath: string;
    name?: string | null;
    mimeType?: string | null;
  }): Promise<Json> {
    const artifactId = newUlid();
    const fallbackName = input.sourcePath.split('/').pop() ?? '';
    const artifactName = (input.name ?? fallbackName).trim();
    if (
      artifactName === '' ||
      artifactName === '.' ||
      artifactName === '..' ||
      artifactName.includes('/') ||
      artifactName.includes('\\')
    ) {
      throw new McpFacadeError('name must be a filename');
    }
    const artifactMime =
      input.mimeType ?? guessMimeType(artifactName) ?? 'application/octet-stream';
    const [record, data] = await this.#call('/internal/mcp/v1/artifacts/submit', input.contextId, {
      artifact_id: artifactId,
      source_path: input.sourcePath,
    });
    const expiresAt = Math.floor(Date.now() / 1000) + this.#settings.artifactTtlSeconds;
    await this.#contextStore.putArtifact(artifactId, {
      name: artifactName,
      mime_type: artifactMime,
      size: Number(data['size']),
      sha256: String(data['sha256']),
      context_id: record.contextId,
      // 存下路径，下载时才能从它借扩展名（displayName 常是没有后缀的标题）。
      path: input.sourcePath,
      // 下载时要拿它做归属维度，见 bridge-client.artifactStream()。
      sandbox_session_id: record.sandboxSessionId,
    });
    const token = this.#signArtifact(artifactId, expiresAt);
    const query = new URLSearchParams({ token }).toString();
    const root = this.#settings.publicBaseUrl.replace(/\/+$/, '');
    const url = `${root}/artifacts/${artifactId}?${query}`;
    return {
      context_id: record.contextId,
      artifact_id: artifactId,
      name: artifactName,
      mime_type: artifactMime,
      size: Number(data['size']),
      sha256: String(data['sha256']),
      download_url: url,
      expires_at: expiresAt,
    };
  }

  async getArtifact(artifactId: string, token: string): Promise<Record<string, unknown> | null> {
    if (!this.verifyArtifactToken(artifactId, token)) return null;
    try {
      return await this.#contextStore.getArtifact(artifactId);
    } catch (error) {
      if (error instanceof ContextStoreError) throw new McpFacadeError(error.message);
      throw error;
    }
  }

  async artifactStream(artifactId: string, sandboxSessionId: string): Promise<Response> {
    try {
      return await this.#bridge.artifactStream(artifactId, sandboxSessionId);
    } catch (error) {
      if (error instanceof SandboxBridgeError) throw new McpFacadeError(error.message);
      throw error;
    }
  }
}

File: sandbox/mcp/service.py
"""Facade business logic: context routing, bridge calls, and artifact URLs."""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import mimetypes
import time
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from pathlib import PurePosixPath
from typing import Any
from urllib.parse import urlencode

from sandbox.mcp.ulid import new_ulid
from sandbox.mcp.context_store import ContextRecord, ContextStore, ContextStoreError
from sandbox.mcp.sandbox_client import SandboxBridgeClient, SandboxBridgeError
from sandbox.mcp.settings import McpSettings


class McpFacadeError(RuntimeError):
    """Safe message returned through the MCP protocol."""


def _b64(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _unb64(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


class McpFacadeService:
    def __init__(
        self,
        settings: McpSettings,
        context_store: ContextStore | None = None,
        bridge: SandboxBridgeClient | None = None,
    ) -> None:
        self.settings = settings
        self.context_store = context_store or ContextStore(settings)
        self.bridge = bridge or SandboxBridgeClient(settings)

    async def start(self) -> None:
        self.settings.validate_runtime()
        await self.context_store.start()
        await self.bridge.start()

    async def close(self) -> None:
        await self.bridge.close()
        await self.context_store.close()

    @staticmethod
    def _context_payload(record: ContextRecord) -> dict[str, str]:
        return {
            "sandbox_session_id": record.sandbox_session_id,
            "workspace_id": record.workspace_id,
        }

    async def _resolve(self, context_id: str | None) -> ContextRecord:
        try:
            return await self.context_store.resolve(
                context_id,
                lambda record: self.bridge.post(
                    "/internal/mcp/v1/context/ensure", self._context_payload(record)
                ),
            )
        except (ContextStoreError, SandboxBridgeError) as exc:
            raise McpFacadeError(str(exc)) from exc

    async def _call(
        self,
        path: str,
        context_id: str | None,
        payload: dict[str, Any],
    ) -> tuple[ContextRecord, dict[str, Any]]:
        record = await self._resolve(context_id)
        try:
            data = await self.bridge.post(path, self._context_payload(record) | payload)
            await self.context_store.touch(record)
            return record, data
        except (ContextStoreError, SandboxBridgeError) as exc:
            raise McpFacadeError(str(exc)) from exc

    async def execute_python(
        self, *, context_id: str | None, code: str, timeout_seconds: int = 120
    ) -> dict[str, Any]:
        if len(code) > self.settings.max_code_length:
            raise McpFacadeError("code exceeds MCP size limit")
        if not 1 <= timeout_seconds <= self.settings.max_timeout_seconds:
            raise McpFacadeError("timeout_seconds exceeds MCP limit")
        record, data = await self._call(
            "/internal/mcp/v1/python/execute",
            context_id,
            {"code": code, "timeout_seconds": timeout_seconds},
        )
        return {"context_id": record.context_id, **data}

    async def execute_shell(
        self, *, context_id: str | None, command: str, timeout_seconds: int = 120
    ) -> dict[str, Any]:
        if len(command) > self.settings.max_command_length:
            raise McpFacadeError("command exceeds MCP size limit")
        if not 1 <= timeout_seconds <= self.settings.max_timeout_seconds:
            raise McpFacadeError("timeout_seconds exceeds MCP limit")
        record, data = await self._call(
            "/internal/mcp/v1/shell/execute",
            context_id,
            {"command": command, "timeout_seconds": timeout_seconds},
        )
        return {"context_id": record.context_id, **data}

    async def file_write(
        self,
        *,
        context_id: str | None,
        path: str,
        content: str,
        mode: str = "overwrite",
    ) -> dict[str, Any]:
        if mode not in {"overwrite", "append"}:
            raise McpFacadeError("mode must be overwrite or append")
        if len(content.encode("utf-8")) > self.settings.max_file_size_bytes:
            raise McpFacadeError("content exceeds MCP file size limit")
        record, data = await self._call(
            "/internal/mcp/v1/files/write",
            context_id,
            {"path": path, "content": content, "mode": mode},
        )
        return {"context_id": record.context_id, **data}

    async def file_read(
        self,
        *,
        context_id: str | None,
        path: str,
        offset: int | None = None,
        limit: int | None = None,
    ) -> dict[str, Any]:
        record, data = await self._call(
            "/internal/mcp/v1/files/read",
            context_id,
            {"path": path, "offset": offset, "limit": limit},
        )
        return {"context_id": record.context_id, **data}

    async def file_list(
        self,
        *,
        context_id: str | None,
        path: str = ".",
        depth: int = 1,
    ) -> dict[str, Any]:
        if not 0 <= depth <= 5:
            raise McpFacadeError("depth must be in 0..5")
        record, data = await self._call(
            "/internal/mcp/v1/files/list",
            context_id,
            {"path": path, "depth": depth},
        )
        return {"context_id": record.context_id, **data}

    def _sign_artifact(self, artifact_id: str, expires_at: int) -> str:
        body = _b64(json.dumps({"artifact_id": artifact_id, "exp": expires_at}, separators=(",", ":")).encode())
        signature = hmac.new(
            self.settings.download_secret.encode("utf-8"), body.encode("ascii"), hashlib.sha256
        ).digest()
        return f"{body}.{_b64(signature)}"

    def verify_artifact_token(self, artifact_id: str, token: str) -> bool:
        try:
            body, signature = token.split(".", 1)
            expected = hmac.new(
                self.settings.download_secret.encode("utf-8"), body.encode("ascii"), hashlib.sha256
            ).digest()
            if not hmac.compare_digest(_unb64(signature), expected):
                return False
            payload = json.loads(_unb64(body))
            return (
                isinstance(payload, dict)
                and payload.get("artifact_id") == artifact_id
                and int(payload.get("exp", 0)) >= int(time.time())
            )
        except (ValueError, TypeError, json.JSONDecodeError, UnicodeDecodeError):
            return False

    async def artifact_submit(
        self,
        *,
        context_id: str | None,
        source_path: str,
        name: str | None = None,
        mime_type: str | None = None,
    ) -> dict[str, Any]:
        artifact_id = new_ulid()
        artifact_name = (name or PurePosixPath(source_path).name).strip()
        if not artifact_name or artifact_name in {".", ".."} or "/" in artifact_name or "\\" in artifact_name:
            raise McpFacadeError("name must be a filename")
        artifact_mime = mime_type or mimetypes.guess_type(artifact_name)[0] or "application/octet-stream"
        record, data = await self._call(
            "/internal/mcp/v1/artifacts/submit",
            context_id,
            {"artifact_id": artifact_id, "source_path": source_path},
        )
        expires_at = int(time.time()) + self.settings.artifact_ttl_seconds
        await self.context_store.put_artifact(
            artifact_id,
            {
                "name": artifact_name,
                "mime_type": artifact_mime,
                "size": int(data["size"]),
                "sha256": str(data["sha256"]),
                "context_id": record.context_id,
            },
        )
        token = self._sign_artifact(artifact_id, expires_at)
        query = urlencode({"token": token})
        root = self.settings.public_base_url.rstrip("/")
        url = f"{root}/artifacts/{artifact_id}?{query}" if root else f"/artifacts/{artifact_id}?{query}"
        return {
            "context_id": record.context_id,
            "artifact_id": artifact_id,
            "name": artifact_name,
            "mime_type": artifact_mime,
            "size": int(data["size"]),
            "sha256": str(data["sha256"]),
            "download_url": url,
            "expires_at": expires_at,
        }

    async def get_artifact(self, artifact_id: str, token: str) -> dict[str, Any] | None:
        if not self.verify_artifact_token(artifact_id, token):
            return None
        try:
            return await self.context_store.get_artifact(artifact_id)
        except ContextStoreError as exc:
            raise McpFacadeError(str(exc)) from exc

    @asynccontextmanager
    async def artifact_stream(self, artifact_id: str) -> AsyncIterator[Any]:
        try:
            async with self.bridge.artifact_stream(artifact_id) as response:
                yield response
        except SandboxBridgeError as exc:
            raise McpFacadeError(str(exc)) from exc


__all__ = ["McpFacadeError", "McpFacadeService"]
